task_02/nikolai_chernyshov/main.cpp

```diff
@@ -0,0 +1,175 @@
+#include <memory>
+#include <cstdlib>
+#include <iostream>
+#include <fstream>
+
+struct edge {
+    static int edge_id_max;
+    int edge_id;
+    int to_vertice;
+    std::shared_ptr<struct vertice> ptr;
+
+    edge() {}
+
+    edge(int to_vertice_, bool again) {
+        if (again) {
+            edge_id = edge_id_max - 1;
+        }
+        else {
+            edge_id = edge_id_max;
+            edge_id_max++;
+        }
+        to_vertice = to_vertice_;
+    }
+};
+
+int edge::edge_id_max = 0;
+
+struct vertice {
+    static int vertice_id_max;
+    int id;
+    int depth;
+    std::unique_ptr<struct edge[]> edges_from_ptr;
+    std::unique_ptr<struct edge> parent_edge_ptr;
+    int edges_from_num;
+
+    //How many new edges to generate from current vertice? 0..3
+    int num_generate() {
+        int res;
+        if (this->depth == 1) return 3;
```

Comment:
Зачем ты используешь `this`?
```suggestion
        if (depth == 1) return 3;
```